### PR TITLE
make ResourceTracker to own cluster-scope resource

### LINKED ISSUE #1614: Can't re-use pv (retain) after deleting vela app
### Discussed in https://github.com/oam-dev/kubevela/discussions/1574

<div type='discussions-op-text'>

<sup>Originally posted by **hprotzek** April 27, 2021</sup>
Hi there, I am trying to create a gcs bucket trait with the `csi-gcs` driver. Using existing buckets, forces me to use `Retain` policy on the pv. When I delete the vela app, the pc get in state `Released` but still have the old owner ref, which prevents using it with a new `vela up ...`

```
cannot apply trait "v1" "PersistentVolume" "csi-gcs-pv-snbucket-test": cannot apply ApplyOption: existing object is not controlled by UID "c4875aba-b301-4072-a388-5cc4e4013be8
```

```
# Code generated by KubeVela templates. DO NOT EDIT.
apiVersion: core.oam.dev/v1beta1
kind: TraitDefinition
metadata:
  name: snbucket
  annotations:
    definition.oam.dev/description: Mount a Google Cloud Storage bucket
spec:
  appliesToWorkloads:
    - snweb
  schematic:
    cue:
      template: |
        parameter: {
        	bucketName: string
        	// +usage=Mount path in container
        	mountPath: string
        	// +usage=Service account json
        	serviceAccount: string
        }
        patch: {
        	spec: template: spec: {
        		// +patchKey=name
        		containers: [{
        			name: context.name
        			// +patchKey=name
        			volumeMounts: [{
        				name:      "bucket1"
        				mountPath: parameter.mountPath
        			}]
        		}]
        		// +patchKey=name
        		volumes: [{
        			name: "bucket1"
        			persistentVolumeClaim: claimName: "csi-gcs-pvc-\(context.name)"
        		}]
        	}
        }
        outputs: {
        	secret: {
        		apiVersion: "v1"
        		kind:       "Secret"
        		metadata: name: "bucket-sa-\(context.name)"
        		type: "Opaque"
        		stringData: {
        			bucket: parameter.bucketName
        			key:    parameter.serviceAccount
        		}
        	}
        
        	pv: {
        		apiVersion: "v1"
        		kind:       "PersistentVolume"
        		metadata: name: "csi-gcs-pv-\(context.name)"
        		spec: {
        			accessModes: ["ReadWriteOnce"]
        			capacity: storage: "5Gi"
        			persistentVolumeReclaimPolicy: "Retain"
        			storageClassName:              "csi-gcs-test-sc"
        			csi: {
        				driver:       "gcs.csi.ofek.dev"
        				volumeHandle: "csi-gcs"
        				nodePublishSecretRef: {
        					name:      "bucket-sa-\(context.name)"
        					namespace: context.namespace
        				}
        			}
        		}
        	}
        
        	pvc: {
        		apiVersion: "v1"
        		kind:       "PersistentVolumeClaim"
        		metadata: name: "csi-gcs-pvc-\(context.name)"
        		spec: {
        			accessModes: ["ReadWriteOnce"]
        			resources: requests: storage: "5Gi"
        			storageClassName: "csi-gcs-test-sc"
        		}
        	}
        }
 ```

Any idea how to solve that? I could generate a random name for the pv, but CUE doesn't provide any random functions, what I see so far? Or could vela remove the owner ref?</div>

### PR DESCRIPTION
**Pull request by captainroy-hy**
to fix #1614 
Signed-off-by: roy wang <seiwy2010@gmail.com>

### CODE REVIEW

**Comment by wangyikewxgm** [comment]
Hi, please try to fix test coverage warning.

**Comment by wangyikewxgm** [comment]
I can't find cluster-scope related garbageCollection logic.If user update application by deleting the cluster-scope trait but resourceTracker still exist, does the trait can be deleted rightly?@captainroy-hy

**Comment by captainroy-hy** [comment]
> I can't find cluster-scope related garbageCollection logic.If user update application by deleting the cluster-scope trait but resourceTracker still exist, does the trait can be deleted rightly?@captainroy-hy

You mean current logic handles gc of resource tracker only for  namespaced resource ?

**Comment by wangyikewxgm** [comment]
> > I can't find cluster-scope related garbageCollection logic.If user update application by deleting the cluster-scope trait but resourceTracker still exist, does the trait can be deleted rightly?@captainroy-hy
> 
> You mean current logic handles gc of resource tracker only for namespaced resource ?

Checked， current  garbageCollection logic could take over gc cluster-scope resource, please add e2e-test for this situation.

File: pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go [hunk 1]
```diff
@@ -347,6 +347,16 @@ func setTraitProperties(t *unstructured.Unstructured, traitName, namespace strin
 	if t.GetName() == "" {
 		t.SetName(traitName)
 	}
+
+	if controller := metav1.GetControllerOf(t); controller != nil {
```
Comment (wangyikewxgm) [inline pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go:351]: Does cluster-scope workload also need this logic?

Comment (captainroy-hy) [reply pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go:351]: Currently, that's the easiest way to judge it's a cluster-scoped resource. I just avoid to involve `discovermapper` pkg method here to keep minimal changes.

Comment (wangyikewxgm) [reply pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go:351]: https://github.com/oam-dev/kubevela/blob/a86ec24ee477db0d9360ce563513065de824a205/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go#L203.  I mean current component render logic would set namespace and ownerRef for workload too. Does this matter for cluster-scope workload?

Comment (captainroy-hy) [reply pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go:351]: The function name is `setTraitProperties`, I think it only works for trait.

Comment (wangyikewxgm) [reply pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/render.go:351]: So, current cannot support cluster-scope workload?

**Review by wonderflow — APPROVED** [review]

**Pull request merged**